library/common/network/configurator.cc

```diff
@@ -167,13 +172,36 @@ void Configurator::reportNetworkUsage(envoy_netconf_t configuration_key, bool ne
 
   // If configuration state changed, refresh dns.
   if (configuration_updated) {
-    refreshDns(configuration_key);
+    refreshDns(configuration_key, false);
+  }
+}
+
+void Configurator::onDnsResolutionComplete(
+    const std::string& host, const Extensions::Common::DynamicForwardProxy::DnsHostInfoSharedPtr&,
+    Network::DnsResolver::ResolutionStatus status) {
+  if (enable_drain_post_dns_refresh_ && pending_drain_) {
+    pending_drain_ = false;
+    if (status == Network::DnsResolver::ResolutionStatus::Success) {
```

Comment:
I'm not sure if we want to gate this on success or not. If we are doing this because of a network switch or background to foreground switch it seems better to just always do it after we tried to refresh DNS?

Can we add more comments somewhere on generally when this is going to fire? Does this change only do this on network switch or are we also doing background to foreground as well?

Also, if we end up leaving this implementation and not doing the per-host version, can you add comments/TODO for the downsides of this implementation? I'm torn on whether it's worth it to do this change or not without the per-host version. It seems not that much harder to me to snap the cache state with all hosts and then just keep track of refresh state for each one.

Reply:
Removed the check, and added an additional comment. I think it may be possible to refine this logic to only drain for a specific host when either DNS has changed, or we have reason to believe prior connections are probably defunct, but after further consideration, I agree that the simplest thing to do is always trigger drain.